Import blog post (#1090)
pulumi/pulumi-hugo: themes/default/content/blog/changes-to-import/index.md

---

> title: "Improving the `pulumi import` experience"

Author: Got any ideas?

> title: "Improving the `pulumi import` experience"

Reviewer: What about "Pulumi Import Improvements: Generate IaC for ANY Cloud Resource"?

> title: "Improving the `pulumi import` experience"

Reviewer: How about "Pulumi Import Improvements: Generate IaC for ANY Cloud Resource"?

At Pulumi, we understand that many cloud engineers and platform teams around the world don't have the luxury of greenfield projects, more often than not we're stuck with the impossible task of "refactoring" or "migrating" existing projects to more modern stacks to help increase team productivity, velocity, and stability. These projects aren't trivial and we want to make it easier for teams and organizations to bring their infrastructure into a cloud engineering world. Oh, and worry not you lucky greenfielders ... even if you just wanna ClickOps your way through some resources and import them into your program; that's gonna work just fine too; we won't tell if you don't 😉

To help you understand the changes we've made to help you on your journey, let's take a look at some side by sides examples of how import used to work vs. what we're releasing today.

### Example: AWS S3 Bucket

> Previously when importing an S3 bucket, we would set the default values for properties like `acl` and `forceDestroy`. The new behavior will not include properties with default values in the generated code, keeping your code more inline with how you'd write it yourself. Artisanal codegen for the artisinal cloud engineer.

Reviewer 2: artisinal or Artisanal?
```suggestion
Previously when importing an S3 bucket, we would set the default values for properties like `acl` and `forceDestroy`. The new behavior will not include properties with default values in the generated code, keeping your code more inline with how you'd write it yourself. Artisanal codegen for the artisanal cloud engineer.
```

#### Code Comparison

This code would have been generated by the older implementation:

```typescript
const my_bucket = new aws.s3.Bucket("my-bucket", {
    acl: "private",
    bucket: "my-bucket-3f85c54",
    forceDestroy: false,
}, {
    protect: true,
});
```

This code will be generated by the newer implementation:

```typescript
const my_bucket = new aws.s3.Bucket("my-bucket", {
    arn: "arn:aws:s3:::my-bucket-3f85c54",
    bucket: "my-bucket-3f85c54",
    hostedZoneId: "Z3BJ6K6RIION7M",
    requestPayer: "BucketOwner",
}, {
    protect: true,

> });

Reviewer 2: Curious why we lead with this example - doesn't it look worse than before? Are there any killer examples when the output actually got shorter?

Author: I don't think any examples got shorter. So many resource have these automatic but not default values.

Reviewer: How about "Pulumi Import Improvements: Generate IaC for ANY Cloud Resource"?

```

The above accurately reflects what the providers `Read` function has told us of the inputs set for this bucket, although due to the way the AWS provider happens to work we could elide all these properties. This is because the provider will use outputs saved in the state file and the new inputs to calculate the update diff, and a property that exists in the old output and is missing from the new inputs is simply left at the old output value. As none of the properties for `Bucket` are marked required the following also works and is equivalent:

```

> const my_bucket = new aws.s3.Bucket("my-bucket", { }, { protect: true });

Reviewer 2: Can it be this

```suggestion
const my_bucket = new aws.s3.Bucket("my-bucket", { protect: true });
```

```

### Example: AWS EC2 Instances

> Previously, when trying to import an EC2 instance with `pulumi import`, you'd be presented with a list of required properties that weren't satisfied. The challenge here is that none of these fields are required by the schema, but at least one is required for the resource to be valid. Like above, the new behavior will read from the provider and inherit the missing properties, providing everything required to codegen. This provides a much cleaner and intuitive experience for the user.

Reviewer 2: This is fairly full with jargon. I don't think most readers know what "schema" or "codegen" are exactly.

#### Old Behavior

```shell
aws:ec2:Instance (test):
  error: aws:ec2/instance:Instance resource 'test' has a problem: Missing required argument: "instance_type": one of `instance_type,launch_template` must be specified. Examine values at 'Instance.InstanceType'.
  error: aws:ec2/instance:Instance resource 'test' has a problem: Missing required argument: "launch_template": one of `ami,instance_type,launch_template` must be specified. Examine values at 'Instance.LaunchTemplate'.
  error: aws:ec2/instance:Instance resource 'test' has a problem: Missing required argument: "ami": one of `ami,launch_template` must be specified. Examine values at 'Instance.Ami'.
```

#### New Behavior

The import will be successful and we'll be presented with a rich resource in your language of choice.

> ```typescript

Reviewer 2: Consider adding examples in other languages?

const test = new aws.ec2.Instance("test", {
    ami: "ami-082b5a644766e0e6f",
    associatePublicIpAddress: true,
    availabilityZone: "us-west-2b",
    capacityReservationSpecification: {
        capacityReservationPreference: "open",
    },
    cpuCoreCount: 1,
    cpuThreadsPerCore: 1,
    creditSpecification: {
        cpuCredits: "standard",
    },

>     iamInstanceProfile: "",

Reviewer 2: Aren't we eliding empty strings now?

Author: We should be, I'll double check this one.

Author: I think its getting confused because this is either a string or an object. We should fix this.

Author: https://github.com/pulumi/pulumi-terraform-bridge/issues/481 So we don't forget

There are two complications with fixing the above. The first is that during import if `Read` doesn't return valid results then `Check` can fail and so we won't have a property set returned from it that we can save to the state file. This case will always have to fall back to just storing the properties returned by `Read` and triggering a check failure during `up` that will need to be manually resolved. The second complication is that the code generation for import works off the state file rather than a separate data flow specifically for imported resources. We plan on improving this but in the spirit of agility we felt that a slightly better import feature today would be more useful to our users than a perfect import feature later.

> We'll go through some examples comparing the old and new system in a bit.

Author: Ah yeh this got moved around.

> ### Terraform `Read`

Author: This was mostly to call out (https://github.com/pulumi/pulumi-terraform-bridge/pull/453), open to a shorter way to explain it.

---

> title: "Changes to `pulumi import`"

Reviewer 3: This could use a more descriptive yet exciting title

Author: Open to suggestions. Some ideas:
Improvements to `pulumi import`
A new import system
Overhaul of the Pulumi import system

# previews. This field is required or the build will fail the linter test.
meta_desc: Changes to the pulumi import system and terraform bridge

# The meta_image appears in social-media previews and on the blog home page.
# A placeholder image representing the recommended format, dimensions and aspect
# ratio has been provided for you.
meta_image: meta.png

# At least one author is required. The values in this list correspond with the `id`
# properties of the team member files at /data/team/team. Create a file for yourself
# if you don't already have one.
authors:
    - fraser-waters

# At least one tag is required. Lowercase, hyphen-delimited is recommended.
tags:
    - features

# See the blogging docs at https://github.com/pulumi/docs/blob/master/BLOGGING.md.
# for additional details, and please remove these comments before submitting for review.
---

Importing resources to Pulumi is a fundamental requirement to allow bringing over existing infrastructure into Pulumi programs. Unfortunately the existing system has resulted in users hitting a number of issues that has made this task harder than it ought to be.

> Today we're announcing plans to rework the import system to try and fix most of these issues. This system is still not perfect and we want to give users a heads up on issues they may still need to work around.

Reviewer 3: This paragraph is a little unclear. Are you saying that we have made changes to how import works, or that we are planning to make changes, or both? It could be clarified in here and also give more of a preview to readers as to what those changes are and how it benefits them.

Author: Well it depends when we push this vs the CLI release. I was thinking we could release this as soon as possible and then actually make the CLI changes the week after. But product team might decided it best to just release them together. A confirmation from @meagancojocar and @lukehoban on what best path to take here would be good.

Reviewer 3: we should be releasing the blog and the update together rather than decoupled. that's the strategy for launches we set last year described in this [doc](https://docs.google.com/document/d/1dGXVjYms3pEqv58V9YSYwOmz0-OD6_toaFQhhz8lKxA/edit)

Reviewer 4: Yeah - let's hold off on the blog post until we've delivered both the CLI feature and the resource provider updates for at least the "big 4" providers.  We can then frame this post as detailing how we've "improved the `pulumi import` experience".

Author: The terraform changes are already landed in tfbridge and more landed today, so their making their way out. The main engine changes are targeted for the March 9th release.

<!--more-->

Pulumi supports two styles of [importing resources]({{< relref "docs/guides/adopting/import/" >}}), using the `pulumi import` command or using the `import` resource option. The changes we're making are mostly for the `pulumi import` command. The resource option behavior will be affected but we don't expect it to have any noticeable user impact.

## The importance of `Provider.Read`

Our import system depends on the ability for a [resource provider]({{< relref "docs/intro/concepts/resources/providers/" >}}) to be able to read the existing state of a resource and report back to Pulumi the current value of it's inputs and outputs.

This isn't always possible to return accurately. For example if a resource's inputs don't match up 1-to-1 with it's output state and the underlying provider can only read the current output state there's no way for it to always construct the correct input state. There could also be bugs in the provider's `Read` method that result in inaccurate reads. We've designed the import system to be tolerant to these cases, but it does mean we don't expect perfect import results for every resource.

> ## The current import system

Reviewer 3: This whole section and the following section provide a lot of info, but it's quite dense to get through before getting to the examples that follow. Do we need to share this much info? I see that it is useful to explain how the existing and new way of importing works so that they can understand the mechanics and also enable them to troubleshoot issues on their own.

Author: We could try flipping things around, show the examples first with maybe some quick explanations of why its changed and then leave the details for afterwards?

Reviewer 5: we will be updating the docs as well right? if so, then possibly consider linking to the docs at the end and reducing the content in the blog.

Author: > we will be updating the docs as well right?

Ah good point they do need updating because the code gen and some defaults have changed, but they don't really talk about how the system works. We wanted to have this blog as a way to help people understand why some imports didn't work, and in the future why they would be getting non-optimal results from an import. I'm not sure how much of that detail we want to include in the day-to-day docs.

Reviewer 5: do yall think that users will care about why they didn't work? i personally think most folks would want to know how it works now, and how it will solve their problem. but i might be missing a bunch of context you have.

Author: That's probably true. I'll give it a think if we can rephrase this just in terms of how the new system works. I still think the examples of "this used to not work, now it does" are valid.

Author: I've flipped the order of the blog around. Examples first, technical details after with an explanation that you don't need to understand it to use import but it might be of interest for understanding why import results look like they do.

---

> title: "Overhaul of the `pulumi import` system"

Reviewer 6: Why "Improving ..." vs something like "Enabling You to Import and Generate IaC for *ANY* Resource"

cc @gthuang @aaronkao 

Let's make it exciting! "Improving", insider baseball, etc, are boring ways to frame what is actually amazing work!

> title: "Overhaul of the `pulumi import` system"

Reviewer 6: Also cc @Brant-Soler on whether he has ideas on title/meta improvements for SEO 😄

> title: "Overhaul of the `pulumi import` system"

Reviewer 4: We could approach this entirely differently as a "re-launch" of the feature.  Basically pretend as though https://www.pulumi.com/blog/pulumi-import-generate-iac-for-existing-cloud-resources/ wasn't written, and write that blog post again relative to the new state of the feature.

Ultimately - there are three options:
1. Just update the existing blog post to match the current reality - and re-promote it.
2. Write a blog post about the improvements we've made (but emphasize the current state of the feature, not the diff)
3. Write a new blog post that re-announces the feature from scratch (effectively ignoring the previous post).

I was suggesting (2), but sounds like Joe sees (3) as a better approach.  Happy to go with that.

> title: "Overhaul of the `pulumi import` system"

Reviewer 7: The link (https://www.pulumi.com/blog/pulumi-import-generate-iac-for-existing-cloud-resources/) has a history with Google, and we can cash in on that. I recommend expanding on the article to:

1. Make it more exciting!
2. Include the past AND current state to focus on growth.
3. Include how we arrived at the current state and HOW that will inform a future state. 
4. Include more code snippets, these are important for catching how developers search for information.

A few questions to ignite creativity:

Is it possible for some developers to use the previous version if they choose? It's popular for certain editions of any piece of software to be chosen over other newer ones like versions of PHP.

Can you think of any external voices that would be willing to contribute? Testimonials are powerful in every form, if you can get a developer talking about why they make a certain choice, code with a certain pattern, or adhere to a certain style guide, other developers can relate to it.

One thing I'm seeing is brevity over depth in many areas of the website, which I completely understand for focus. I'm a fan of deep knowledge and extensive human-spoken English (transcripts for all videos should be available beneath the imported video tucked within an accordion) because of my SEO perspective.

With the new changes an EC2 instance can be imported without warnings:

```
const test = new aws.ec2.Instance("test", {
    ami: "ami-082b5a644766e0e6f",
    associatePublicIpAddress: true,
    availabilityZone: "us-west-2b",
    capacityReservationSpecification: {
        capacityReservationPreference: "open",
    },
    cpuCoreCount: 1,
    cpuThreadsPerCore: 1,
    creditSpecification: {
        cpuCredits: "standard",
    },
    iamInstanceProfile: "",
    instanceInitiatedShutdownBehavior: "stop",
    instanceType: "t2.micro",
    metadataOptions: {
        httpEndpoint: "enabled",
        httpPutResponseHopLimit: 1,
        httpTokens: "optional",
    },
    privateIp: "172.31.20.37",
    rootBlockDevice: {
        iops: 100,
        volumeSize: 8,
        volumeType: "gp2",
    },
    securityGroups: ["secgrp-b0c0140"],
    subnetId: "subnet-c7d926bf",
    tenancy: "default",
    vpcSecurityGroupIds: ["sg-053fffb857c2a2eec"],
}, {
    protect: true,
});
```

> Again this reflects what `Read` has told us of the inputs set for this instance, and again due to the way the AWS provider works many of these inputs could be elided and the provider would pick up the values from the saved output set. The following is again equivalent to the above given the saved state:

Author: Yeh this is actually showing what the state of your instance is. A lot of these are kinda of default values but they're not expressed that way in the schema. Like rootBlockDevice isn't something you'd normally need to specify, I made this instance to import with something like:
```
const instance = new aws.ec2.Instance("my-instance", {
    ami: "ami-082b5a644766e0e6f",
    instanceType: "t2.micro",
    tags: {
        purpose: "testing",
    },
});
```
but when reading it back there's no way to tell from the aws call if volumeSize and the like where set by the provider to default values or if they were originally specified.

I'm not sure what the best way to express this "more accurate, but maybe more than you need" state of things.

---
title: "Improving the `pulumi import` experience"
date: 2022-03-11
meta_desc: Recent improvements to `pulumi import` make it an even smoother way to import existing cloud infrastructure into Pulumi.
meta_image: meta.png
authors:
    - fraser-waters
    - david-flanagan
tags:
    - features
    - migration
    - import
---

> Last year, we [introduced a new feature]({{< relref "blog/pulumi-import-generate-iac-for-existing-cloud-resources" >}}) for Pulumi that allowed you to import existing infrastructure into your Pulumi program. Not only did it bring the resource into the Pulumi state file, but it could generate the source code for your Pulumi program too. Today, we're excited to announce that we've listened to feedback and delivered a plethora of updates an fixes to streamline the import experience; to make it **more useful**, **more convenient**, and **more powerful**.

Reviewer: ```suggestion
Last year, we [introduced a new Pulumi feature]({{< relref "blog/pulumi-import-generate-iac-for-existing-cloud-resources" >}}) that allows you to import existing infrastructure into your Pulumi program. Not only did it bring the resource into the Pulumi state file, but it could generate the source code for your Pulumi program too. Today, we're excited to announce that we've listened to feedback and delivered a plethora of updates and fixes to streamline the import experience; to make it **more useful**, **more convenient**, and **more powerful**.
```

> At Pulumi, we understand that many cloud engineers and platform teams around the world don't have the luxury of greenfield projects, more often than not we're stuck with the impossible task of "refactoring" or "migrating" existing projects to more modern stacks to help increase team productivity, velocity, and stability. These projects aren't trivial and we want to make it easier for teams and organizations to bring their infrastructure into a cloud engineering world. Oh, and worry not you lucky greenfielders ... even if you just wanna ClickOps your way through some resources and import them into your program; that's gonna work just fine too; we won't tell if you don't 😉

Reviewer: ```suggestion
At Pulumi, we understand that many cloud engineers and platform teams around the world don't have the luxury of greenfield projects, more often than not we're stuck with the impossible task of "refactoring" or "migrating" existing projects to more modern stacks to help increase team productivity, velocity, and stability. These projects aren't trivial and we want to make it easier for teams and organizations to bring their infrastructure into a cloud engineering world. Oh, and worry not you lucky greenfielders ... even if you just wanna ClickOps your way through some resources and import them into your program; that's gonna work just fine too; we won't tell if you don't 😉.
```

> To help you understand the changes we've made to help you on your journey, let's take a look at some side by sides examples of how import used to work vs. what we're releasing today.

Reviewer: ```suggestion
To help you understand the changes we've made to help you on your journey, let's take a look at some side-by-side examples of how `pulumi import` used to work vs. what we're releasing today.
```

### Example: AWS S3 Bucket

> Previously when importing an S3 bucket, we would set the default values for properties like `acl` and `forceDestroy`. The new behavior will not include properties with default values in the generated code, keeping your code more inline with how you'd write it yourself. Artisanal codegen for the artisanal cloud engineer.

Reviewer: ```suggestion
Previously when importing an S3 bucket, we would set the default values for properties like `acl` and `forceDestroy`. The new behavior will not include properties with default values in the generated code, keeping your code inline with how you'd write it yourself. Artisanal codegen for the artisanal cloud engineer.
```

#### Code Comparison

This code would have been generated by the older implementation:

```typescript
const my_bucket = new aws.s3.Bucket("my-bucket", {
    acl: "private",
    bucket: "my-bucket-3f85c54",
    forceDestroy: false,
}, {
    protect: true,
});
```

This code will be generated by the newer implementation:

```typescript
const my_bucket = new aws.s3.Bucket("my-bucket", {
    arn: "arn:aws:s3:::my-bucket-3f85c54",
    bucket: "my-bucket-3f85c54",
    hostedZoneId: "Z3BJ6K6RIION7M",
    requestPayer: "BucketOwner",
}, {
    protect: true,
});
```

> The above accurately reflects what the providers `Read` function has told us of the inputs set for this bucket, although due to the way the AWS provider happens to work we could elide all these properties. This is because the provider will use outputs saved in the state file and the new inputs to calculate the update diff, and a property that exists in the old output and is missing from the new inputs is simply left at the old output value. As none of the properties for `Bucket` are marked required the following also works and is equivalent:

Reviewer: ```suggestion
The above accurately reflects what the providers `Read` function has told us of the inputs set for this bucket, although due to the way the AWS provider works, we could elide all these properties. This is because the provider will use outputs saved in the state file and the new inputs to calculate the update diff. A property that exists in the old output and is missing from the new inputs is simply left at the old output value. As none of the properties for `Bucket` are marked required the following works as well and is equivalent:
```

```
const my_bucket = new aws.s3.Bucket("my-bucket", { protect: true });
```

### Example: AWS EC2 Instances

> Previously, when trying to import an EC2 instance with `pulumi import`, you'd be presented with a list of required properties that weren't satisfied. The challenge here is that none of these fields are marked as required, but at least one is required for the resource to be valid. Like above, the new behavior will read from the provider and inherit the missing properties, providing everything required to correctly import. This provides a much cleaner and intuitive experience for the user.

Reviewer: ```suggestion
Previously, when trying to import an EC2 instance with `pulumi import`, you'd be presented with a list of required properties that weren't satisfied. Like above, the new behavior will read from the provider and inherit the missing properties, providing everything required to correctly import. This provides a cleaner and more intuitive experience for the user.
```

```typescript
const test = new aws.ec2.Instance("test", {
    ami: "ami-082b5a644766e0e6f",
    instanceType: "t2.micro"
}, {
    protect: true,
});
```

### Everything is Better, Yes?

To quote our old friend, Harlan, "Everything is better, yes?"

![Comtrya](./Comtrya.gif)

While we've only shown you a couple of examples of this new behaviour, rest assured that the benefits of it stretch far and wide across the majority of Pulumi resources.

If you've ever had a problem importing a resource before, we encourage you to try it again and let us know if you run into any problems.

Curious how this all works? Let's dive in.

## Technical details

We've never really explained how the import system works before, which has lead to a lot of confusion from users when they encountered errors with it. Given the changes we're making here we felt it a good time to give a more detailed write up of how the engine works to try and help users understand why their import results are the way they are.

> This section is technically detailed and requires some understanding of the Pulumi architecture, but is not necessary to understand to be able to use the import system.

Reviewer: ```suggestion
This section requires some understanding of the Pulumi architecture, but that understanding is not necessary to be able to use `pulumi import`.
```

    "bucket": "my-bucket-3f85c54",
    "corsRules": [],
    "grants": [],
    "hostedZoneId": "Z3BJ6K6RIION7M",
    "lifecycleRules": [],
    "loggings": [],
    "objectLockConfiguration": null,
    "replicationConfiguration": null,
    "requestPayer": "BucketOwner",
    "serverSideEncryptionConfiguration": null,
    "tags": {
        "__defaults": []
    },
    "versioning": {
        "__defaults": [],
        "enabled": false,
        "mfaDelete": false
    },
    "website": null
}
```

We've made [a change](https://github.com/pulumi/pulumi-terraform-bridge/pull/453) to the Terraform bridge to try and elide these zero properties that _should_ just be missing. Reading the same S3 bucket with these changes gives the following much smaller property set:

```
{
    "__defaults": [],
    "arn": "arn:aws:s3:::my-bucket-3f85c54",
    "bucket": "my-bucket-3f85c54",
    "hostedZoneId": "Z3BJ6K6RIION7M",
    "requestPayer": "BucketOwner",
    "website": null
}
```

We'll continue to work on the Terraform bridge to give better results here. For example the `website` property on the bucket should probably be missing rather than null.

## Feedback

> We're aware this could have a large impact to some of our users so we wanted to make sure we explained these changes well. If you have questions or worries about this change let us know at our community [Slack](https://slack.pulumi.com/) or [GitHub Discussions](https://github.com/pulumi/pulumi/discussions).

Reviewer: ```suggestion
We're aware this could have a large impact to some of our users so we wanted to make sure we explained these changes well. If you have questions or concerns about this change let us know in our community [Slack](https://slack.pulumi.com/) or [in GitHub Discussions](https://github.com/pulumi/pulumi/discussions).
```
